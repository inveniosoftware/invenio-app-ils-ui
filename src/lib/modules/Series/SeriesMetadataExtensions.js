import PropTypes from 'prop-types';
import React from 'react';
import { Divider, Table } from 'semantic-ui-react';
import _keys from 'lodash/keys';
import { extensionsConfig } from '@config';

export const SeriesMetadataExtensions = ({ extensions, showDivider }) => {
  return (
    <React.Fragment>
      {showDivider && (
        <Divider horizontal>{extensionsConfig.series.label}</Divider>
      )}
      <Table definition>
        <Table.Body>
          {_keys(extensions).map(key => (
            <Table.Row key={key}>
              <Table.Cell width={4}>
                {extensionsConfig.series.fields[key].label}
              </Table.Cell>
              <Table.Cell>{extensions[key]}</Table.Cell>
            </Table.Row>
          ))}
        </Table.Body>
      </Table>
    </React.Fragment>
  );
};

SeriesMetadataExtensions.propTypes = {
  extensions: PropTypes.object.isRequired,
  showDivider: PropTypes.bool,
};

SeriesMetadataExtensions.defaultProps = {
  showDivider: true,
};
