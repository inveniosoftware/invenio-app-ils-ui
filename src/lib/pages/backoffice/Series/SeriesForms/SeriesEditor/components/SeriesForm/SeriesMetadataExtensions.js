import { invenioConfig } from '@config';
import { AccordionField } from '@forms/core/AccordionField';
import { BooleanField } from '@forms/core/BooleanField';
import { DateInputField } from '@forms/core/DateTimeFields/DateInputField';
import { StringField } from '@forms/core/StringField';
import _merge from 'lodash/merge';
import _forOwn from 'lodash/forOwn';
import _get from 'lodash/get';
import _keys from 'lodash/keys';
import PropTypes from 'prop-types';
import React from 'react';

const getFormComponent = (fieldType) => {
  switch (fieldType) {
    case 'string':
      return StringField;
    case 'boolean':
      return BooleanField;
    case 'date':
      return DateInputField;
    default:
      throw new Error(
        `${fieldType} did not match any of the supported types: string, date or boolean`
      );
  }
};

export const SeriesMetadataExtensions = ({ extensions }) => {
  const { label, fields } = invenioConfig.SERIES.extensions;
  const configDefaults = {};
  _keys(fields).map((key) => (configDefaults[key] = fields[key]['default']));
  const allExtensions = _merge(configDefaults, extensions);

  let components = [];
  _forOwn(allExtensions, (value, key) => {
    const componentType = fields[key].type;
    const Component = getFormComponent(componentType);

    components.push(
      <Component
        fieldPath={`extensions.${key}`}
        key={key}
        label={key}
        optimized
        required={_get(fields, `${key}.isRequired`, false)}
      />
    );
  });

  return (
    <AccordionField
      label={label}
      fieldPath="extensions"
      content={<>{components.map((cmp) => cmp)}</>}
    />
  );
};

SeriesMetadataExtensions.propTypes = {
  extensions: PropTypes.object.isRequired,
};
