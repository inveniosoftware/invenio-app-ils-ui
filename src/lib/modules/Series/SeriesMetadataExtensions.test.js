import React from 'react';
import { mount } from 'enzyme';
import { SeriesMetadataExtensions } from './SeriesMetadataExtensions';

jest.mock('@config/extensionsConfig');

describe('SeriesMetadataExtensions tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const extensions = {
    'fooObject:fooProp': 'fooValue',
    'barObject:barProp': 'barValue',
  };

  it('should render the SeriesMetadataExtensions', () => {
    component = mount(<SeriesMetadataExtensions extensions={extensions} />);
    expect(component).toMatchSnapshot();
    const rows = component.find('SeriesMetadataExtensions').find('TableRow');
    expect(rows).toHaveLength(2);
  });
});
