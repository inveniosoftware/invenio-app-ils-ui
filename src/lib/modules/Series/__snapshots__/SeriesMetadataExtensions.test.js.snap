// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesMetadataExtensions tests should render the SeriesMetadataExtensions 1`] = `
<SeriesMetadataExtensions
  extensions={
    Object {
      "barObject:barProp": "barValue",
      "fooObject:fooProp": "fooValue",
    }
  }
  showDivider={true}
>
  <Divider
    horizontal={true}
  >
    <div
      className="ui horizontal divider"
    >
      series
    </div>
  </Divider>
  <Table
    as="table"
    definition={true}
  >
    <table
      className="ui definition table"
    >
      <TableBody
        as="tbody"
      >
        <tbody
          className=""
        >
          <TableRow
            as="tr"
            cellAs="td"
            key="fooObject:fooProp"
          >
            <tr
              className=""
            >
              <TableCell
                as="td"
                width={4}
              >
                <td
                  className="four wide"
                >
                  Foo
                </td>
              </TableCell>
              <TableCell
                as="td"
              >
                <td
                  className=""
                >
                  fooValue
                </td>
              </TableCell>
            </tr>
          </TableRow>
          <TableRow
            as="tr"
            cellAs="td"
            key="barObject:barProp"
          >
            <tr
              className=""
            >
              <TableCell
                as="td"
                width={4}
              >
                <td
                  className="four wide"
                >
                  Bar
                </td>
              </TableCell>
              <TableCell
                as="td"
              >
                <td
                  className=""
                >
                  barValue
                </td>
              </TableCell>
            </tr>
          </TableRow>
        </tbody>
      </TableBody>
    </table>
  </Table>
</SeriesMetadataExtensions>
`;
